core/shared/src/main/scala/zio/Fiber.scala

```diff
@@ -757,11 +761,25 @@ object Fiber extends FiberPlatformSpecific {
   def unsafeCurrentFiber(): Option[Fiber[Any, Any]] =
     Option(_currentFiber.get)
 
-  private[zio] def newFiberId(): Fiber.Id = Fiber.Id(System.currentTimeMillis(), _fiberCounter.getAndIncrement())
+  private[zio] val _currentFiber: ThreadLocal[internal.FiberContext[_, _]] =
+    new ThreadLocal[internal.FiberContext[_, _]]()
 
-  private[zio] def untrack[E, A](context: internal.FiberContext[E, A]): Boolean =
-    if (context ne null) Fiber.rootFibers.remove(context)
-    else false
+  private[zio] def get(fiberId: Fiber.Id): UIO[Option[Fiber.Runtime[Any, Any]]] = {
+
+    def loop(fiber: Fiber.Runtime[Any, Any]): UIO[Option[Fiber.Runtime[Any, Any]]] =
+      fiber.id.flatMap { id =>
+        if (id == fiberId) UIO.succeedNow(Some(fiber))
+        else fiber.children.flatMap(ZIO.collectFirst(_)(loop))
+      }
+
+    roots.flatMap(ZIO.collectFirst(_)(loop))
```

Comment:
What happens if there are a million fibers? Or is the idea this is only intended to be used in a test with a small number of fibers?

Reply:
I think most of the time there are only a small number of effects in a test that is using time but yes there is some inefficiency here. One of the things I was wondering about is whether we could create a weak map from fiber IDs to fibers when we create them so we could just look up the fiber ID in the map instead of having to do a search for it. But there is potentially a performance implication from that. We could also potentially memoize the results so only have to search for a given fiber once.